File: 2022/07/19/java/FindDifference.java
/**
 * https://leetcode.com/problems/find-the-difference/
 * @author Andrew Jarombek
 * @since 7/19/2022
 */

import java.util.HashMap;
import java.util.Map;

public class FindDifference {
    public static char findTheDifference(String s, String t) {
        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            int val = map.getOrDefault(c, 0);

            if (val == 0) {
                return c;
            }

            map.put(c, val - 1);
        }

        return ' ';
    }

    public static void main(String[] args) {
        assert findTheDifference("ab", "abc") == 'c';
    }
}

File: 2022/08/02/find_difference.py
"""
https://leetcode.com/problems/find-the-difference-of-two-arrays/
Author: Andrew Jarombek
Date: 8/2/2022
"""

from typing import List, Set


def find_difference(nums1: List[int], nums2: List[int]) -> List[Set[int]]:
    set1 = set(nums1)
    set2 = set(nums2)

    return [set1 - set2, set2 - set1]


if __name__ == '__main__':
    result = find_difference([1, 2, 3, 3], [1, 1, 2, 2])
    assert result[0] == {3}
    assert len(result[1]) == 0
